sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/AnalysisHelper.scala

```diff
@@ -229,6 +229,14 @@ trait AnalysisHelper extends QueryPlan[LogicalPlan] { self: LogicalPlan =>
     resolveExpressionsWithPruning(AlwaysProcess.fn, UnknownRuleId)(r)
   }
 
+  /**
+   * Recursively top-down transforms the expressions of a tree, skipping nodes that have already
+   * been analyzed.
+   */
+  def resolveExpressionsDown(r: PartialFunction[Expression, Expression]): LogicalPlan = {
```

Comment:
super nit: shall we put this method right before `resolveExpressionsDownWithPruning`?

Reply:
done